src/SDKs/Azure.Base/data-plane/Azure.Base/UriExtensions.cs

```diff
@@ -2,22 +2,207 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Text;
 
 namespace Azure
 {
-    public static class UriExtensions
+
+    public class HttpPipelineUriBuilder
     {
-        public static void AppendQuery(this UriBuilder builder, string name, string value)
+        private const char QuerySeperator = '?';
+
+        private readonly StringBuilder _pathAndQuery = new StringBuilder();
+
+        private int _queryIndex = -1;
+
+        private Uri _uri;
+
+        private int _port;
+
+        private string _host;
+
+        private string _scheme;
+
+        public string Scheme
+        {
+            get => _scheme;
+            set
+            {
+                ResetUri();
+                _scheme = value;
+            }
+        }
+
+        public string Host
+        {
+            get => _host;
+            set
+            {
+                ResetUri();
+                _host = value;
+            }
+        }
+
+        public int Port
         {
-            if(!string.IsNullOrEmpty(builder.Query)) {
-                builder.Query = builder.Query + "&" + name + "=" + value;
-            } 
-            else {
-                builder.Query = name + "=" + value;
+            get => _port;
+            set
+            {
+                ResetUri();
+                _port = value;
             }
         }
 
-        public static void AppendQuery(this UriBuilder builder, string name, long value)
-            => AppendQuery(builder, name, value.ToString());
+        public string Query
+        {
+            get => HasQuery ? _pathAndQuery.ToString(_queryIndex, _pathAndQuery.Length - _queryIndex) : string.Empty;
+            set
+            {
+                ResetUri();
+                if (HasQuery)
+                {
+                    _pathAndQuery.Remove(_queryIndex, _pathAndQuery.Length - _queryIndex);
+                    _queryIndex = -1;
+                }
+
+                if (!string.IsNullOrEmpty(value))
```

Comment:
Should we consider an `IsNullOrWhitespace` here?

Reply:
No, I don't think so, ` ` is a valid value for query and path